tcp_sink.go

```diff
@@ -70,27 +76,26 @@ func (w *sinkWriter) Write(p []byte) (int, error) {
 }
 
 func (s *tcpStatsdSink) Flush() {
-	now := time.Now()
-	if err := s.flush(); err != nil {
-		// Not much we can do here; we don't know how/why we failed.
-		return
+	if s.flush() != nil {
+		return // nothing we can do
 	}
+
+	s.doFlush <- struct{}{}
 	s.mu.Lock()
-	for now.After(s.lastFlushTime) {
+	for len(s.outc) != 0 {
```

Comment:
This seems wrong to me and will lead to infinite blocking. I'm currently investigating a case of that in https://jira.lyft.net/browse/DS-6616 which I believe is due to this.

I understand that line 83 is trying to signal to the loop that it should only flush pending (https://github.com/lyft/gostats/pull/79/files#diff-fe0a6341412533cc5c3ed899e22b8610R193), but the condition in the `for` loop will remain true if there is additional items in the channel to flush. As a result, it will go back into the condition wait and never get unblocked thus causing stats to stop flowing (for counters and gauges, though not timers because they use a different path of directly writing to the sink instead of being written in the background loop (https://github.com/lyft/gostats/blob/03ea723b108b9d1d085a9e857fa65027792f37ea/stats.go#L306)).

Reply:
Good catch this can definitely hang if item are added to `outc` while in the flush loop.  Fixed with: https://github.com/lyft/gostats/pull/84.